packages/back-end/src/jobs/updateExperimentResults.ts

```diff
@@ -106,132 +104,184 @@ async function updateSingleExperiment(job: UpdateSingleExpJob) {
 
   if (!experimentId || !orgId) return;
 
-  const context = await getContextForAgendaJobByOrgId(orgId);
-
-  const { org: organization } = context;
-
-  const experiment = await getExperimentById(context, experimentId);
-  if (!experiment) return;
-
-  let project = null;
-  if (experiment.project) {
-    project = await context.models.projects.getById(experiment.project);
+  // This job may take a while to run,
+  // so to ensure another server doesn't pick it up we
+  // Start a timer to call job.touch() every 3 minutes which will update the job's lock time.
+  const TOUCH_INTERVAL = 3 * 60 * 1000;
+  let touchTimer: NodeJS.Timeout | null = null;
+  let finished = false;
+
+  function startTouch() {
+    touchTimer = setInterval(() => {
+      if (!finished) {
+        job.touch().catch((e) => {
+          logger.error(e, "Failed to touch Agenda job");
+        });
+      }
+    }, TOUCH_INTERVAL);
   }
-  const { settings: scopedSettings } = getScopedSettings({
-    organization: context.org,
-    project: project ?? undefined,
-  });
 
-  // Disable auto snapshots for the experiment so it doesn't keep trying to update if schedule is off (non-bandits only)
-  if (
-    organization?.settings?.updateSchedule?.type === "never" &&
-    experiment.type !== "multi-armed-bandit"
-  ) {
-    await updateExperiment({
-      context,
-      experiment,
-      changes: {
-        autoSnapshots: false,
-      },
-    });
-    return;
+  function stopTouch() {
+    finished = true;
+    if (touchTimer) clearInterval(touchTimer);
   }
 
-  try {
-    logger.info("Start Refreshing Results for experiment " + experimentId);
-    const datasource = await getDataSourceById(
-      context,
-      experiment.datasource || ""
-    );
-    if (!datasource) {
-      throw new Error("Error refreshing experiment, could not find datasource");
-    }
-
-    const {
-      regressionAdjustmentEnabled,
-      settingsForSnapshotMetrics,
-    } = await getSettingsForSnapshotMetrics(context, experiment);
-
-    const analysisSettings = getDefaultExperimentAnalysisSettings(
-      experiment.statsEngine || scopedSettings.statsEngine.value,
-      experiment,
-      organization,
-      regressionAdjustmentEnabled
-    );
-
-    const metricMap = await getMetricMap(context);
-    const factTableMap = await getFactTableMap(context);
-
-    let reweight =
-      experiment.type === "multi-armed-bandit" &&
-      experiment.banditStage === "exploit";
-
-    if (experiment.type === "multi-armed-bandit" && !reweight) {
-      // Quick check to see if we're about to enter "exploit" stage and will need to reweight
-      const tempChanges = updateExperimentBanditSettings({
-        experiment,
-        isScheduled: true,
-      });
-      if (tempChanges.banditStage === "exploit") {
-        reweight = true;
-      }
-    }
+  startTouch();
+
+  const TIMEOUT = MAX_QUERY_TIMEOUT_MS + 1000; // Allow some buffer for the query client to close properly after it timesout
```

Comment:
This job will do many Mongo lookups, then run multiple queries, and then call the stats engine.  Depending on concurrency settings, queries may run sequentially as well.  So the job timeout needs to be significantly longer than the individual query timeout.

Maybe it's 1 hour max per query and 2 hours max per job (or something like that).  We could make a new env var if we want for the job timeout.